pyiceberg/utils/singleton.py

```diff
@@ -47,3 +47,7 @@ def __new__(cls, *args, **kwargs):  # type: ignore
         if key not in cls._instances:
             cls._instances[key] = super().__new__(cls)
         return cls._instances[key]
+
+    def __deepcopy__(self, memo: Dict[int, Any]) -> Any:
```

Comment:
nit: what do you think about adding a snippet here in the comments to talk about the motivation for overriding the `__deepcopy__` function here so that future readers don't need to dig through the commit history